lib/ramble/ramble/application.py

```diff
@@ -865,17 +875,97 @@ def _make_experiments(self, workspace):
         experiment_run_dir = self.expander.experiment_run_dir
         fs.mkdirp(experiment_run_dir)
 
-        for template_name, template_val in workspace.all_templates():
+        for template_name, template_conf in workspace.all_templates():
             expand_path = os.path.join(experiment_run_dir, template_name)
 
             with open(expand_path, 'w+') as f:
-                f.write(self.expander.expand_var(template_val))
+                f.write(self.expander.expand_var(template_conf['contents']))
             os.chmod(expand_path, stat.S_IRWXU | stat.S_IRWXG
                      | stat.S_IROTH | stat.S_IXOTH)
 
         experiment_script = workspace.experiments_script
         experiment_script.write(self.expander.expand_var('{batch_submit}\n'))
 
+    def _clean_hash_variables(self, workspace, variables):
+        """Cleanup variables to hash before computing the hash
+
+        Perform some general cleanup operations on variables
+        before hashing, to help give useful hashes.
+        """
+
+        # Purge workspace name, as this shouldn't affect the experiments
+        if 'workspace_name' in variables:
+            del variables['workspace_name']
+
+        # Remove the workspace path from variable definitions before hashing
+        for var in variables:
+            if isinstance(variables[var], six.string_types):
```

Comment:
I'm a bit uneasy about the hashing actually changing the workspace (not that it's actually an issue or its not repeatable)

Is it perhaps better to either: a) clone and edit the attr dict at hash time; or b) separate out the things like root and subpath so we only delete not edit for hashing?

Reply:
What is actually implemented is, I think, basically your (a): https://github.com/GoogleCloudPlatform/ramble/pull/182/files/155880a645977e10222d4f29e4bfc1cfa7c26cb6#diff-f442a3694bd08c70f90d7995078d7a5b62e2443faa816e9f97ada8389dee66c5R916-R917

Also, I wanted to edit rather than delete, since users could have paths in variables that they defined, instead of only Ramble managed ones (which would basically just re-create this issue).